torchgeo/datasets/geo.py

```diff
@@ -307,12 +307,16 @@ def files(self) -> list[Path]:
         # Using set to remove any duplicates if directories are overlapping
         files: set[Path] = set()
         for path in paths:
-            if os.path.isdir(path):
-                pathname = os.path.join(path, '**', self.filename_glob)
-                files |= set(glob.iglob(pathname, recursive=True))
-            elif os.path.isfile(path) or path_is_vsi(path):
-                files.add(path)
-            elif not hasattr(self, 'download'):
+            file_not_found = True
+            if os.path.isfile(path):
+                if fnmatch.fnmatch(str(path), os.path.join('*', self.filename_glob)):
+                    files.add(path)
+                    file_not_found = False
+            elif files_found := set(list_directory_recursive(path, self.filename_glob)):
+                files |= files_found
+                file_not_found = False
+
+            if file_not_found and not hasattr(self, 'download'):
                 warnings.warn(
                     f"Could not find any relevant files for provided path '{path}'. "
                     f'Path was ignored.',
```

Comment:
@adamjstewart I went ahead and prepared how this will look adopting changes from #2222. I think this is the best I can do in terms of clean code.

Reply:
Merged from main with #2222 changes